src/lib/internal_tracing/internal_tracing.mli

```diff
@@ -0,0 +1,158 @@
+          ~transactions ~get_completed_work ~logger ~log_block_creation
+          ~winner_pk:winner_pubkey ~block_reward_threshold
+      in
+      [%log internal] "Generate_next_state_done" ;
+    ]}
+
+    The first parameter should be a checkpoint tag (expected format is ["Name_of_checkpoint"]), and
+    optionally, metadata can be provided too.
+
+    In addition to checkpoints, control commands are also accepted ["@control_command"]:
+
+    {[
+      [%log internal] "@block_metadata"
+        ~metadata:
+          [ ( "blockchain_length"
+            , Mina_numbers.Length.to_yojson
+              @@ Mina_block.blockchain_length
+              @@ Breadcrumb.block breadcrumb )
+          ] ;
+    ]}
+
+    {1 Control commands}
+
+    The current list of control commands that can be issued is:
+
+     - ["@metadata"]: associates metadata to the last produced checkpoint
+       (this has the same result as including a [metadata] parameter when
+       logging a checkpoint).
+     - ["@block_metadata"]: associates metadata to the block currently being
+       processed/produced.
+     - ["@produced_block_state_hash"]: must be issued when the state hash of a
+       produced block is known.
+
+    Internally, these control commands are issued too:
+
+     - ["@current_block"]: used to notify of a execution context change that
+       brings a different block into context.
+     - ["@current_call_id"]: used to notify of a execution context change that
+       brings a different concurrent verifier or prover call into context.
+     - ["@internal_tracing_enabled"]: issued whenever internal tracing is enabled.
+     - ["@internal_tracing_disabled"]: issued whenever internal tracing is disabled.
+     - ["@mina_node_metadata"]: associates global metadata about the current node
+       (useful for version, branch, etc)
+     - ["@rotated_log_end"]: issued when a log file is rotated to notify
+       log consumers.
+     - ["@rotated_log_started"]: issued when a log file is rotated to notify
+       log consumers.
+
+    {1 Output Format}
+
+    The output format is {{: https://jsonlines.org/} JSON Lines}.
+
+    Checkpoints are represented as arrays of two elements, the tag and a timestamp:
+
+    Control commands are represented as JSON objects, with one key and one value.
+    The key is the control command name, and the value is the parameter.contents
+
+    {2 Example}
+
+    {[
+     {"current_block":"3NKk9rxqKp2fyt9fX4JyTu6yRMDfk3k7Wke2aDZVvQs7ReDazxbZ"}
+     ["External_block_received",1677161355.688698]
+     {"block_metadata":{"blockchain_length":"2"}}
+    ]}
+
+    *)
+
+(** {1 API} *)
+
+(** [is_enabled ()] returns [true] if internal tracing is enabled, and [false] otherwise. *)
+val is_enabled : unit -> bool
+
+(** [register_toggle_callback callback] will register [callback] to be called whenever
+    internal tracing is toggled.
+
+    This is useful to syncronize internal tracing done by subprocesses like the verifier
```

Comment:
```suggestion
    This is useful to synchronize internal tracing done by subprocesses like the verifier
```

Reply:
Thanks, I will correct this in https://github.com/MinaProtocol/mina/pull/12874